aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1155,6 +1170,100 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
   return std::tuple<Tensor&, Tensor&>(vals, vecs);
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ eig ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
+static void apply_eig(Tensor& vals_, Tensor& vecs_, const Tensor& self, bool eigenvectors, int64_t* info_ptr) {
+#ifndef USE_LAPACK
+  AT_ERROR("symeig: LAPACK library not found in compilation");
+#else
+  int64_t n = self.size(-1);
+  auto vals_data = vals_.data_ptr<scalar_t>();
+  scalar_t* wr = vals_data;
+  scalar_t* wi = vals_data + n;
+  char jobvr = eigenvectors ? 'V' : 'N';
+  int lda = n;
+  int info;
+  if (n > 0) {
+    scalar_t wkopt;
+    scalar_t* vecs_data = eigenvectors ? vecs_.data_ptr<scalar_t>() : nullptr;
+    int ldvr = eigenvectors ? n : 1;
+
+    lapackEig<scalar_t>('N', jobvr, n, self.data_ptr<scalar_t>(), lda, wr, wi,
+      nullptr, 1, vecs_data, ldvr, &wkopt, -1, &info);
+
+    int lwork = (int)wkopt;
+    Tensor work = at::empty({lwork}, self.dtype());
+
+    lapackEig<scalar_t>('N', jobvr, n, self.data_ptr<scalar_t>(), lda, wr, wi,
+      nullptr, 1, vecs_data, ldvr, work.data<scalar_t>(), lwork, &info);
+    *info_ptr = info;
+  }
+#endif
+}
+
+static std::tuple<Tensor, Tensor> _eig_helper_cpu(const Tensor& self, bool eigenvectors) {
+  TORCH_CHECK(self.dim() == 2, "A should be 2 dimensional");
+  TORCH_CHECK(self.size(0) == self.size(1), "A should be square");
+  TORCH_CHECK(self.isfinite().all().item<bool>(), "A should not contain infs or NaNs");
+  int64_t n = self.size(-1);
+  Tensor self_ = at::empty_strided(
+      {n, n},
+      {1, n},
+      at::TensorOptions(self.dtype()));
+  self_.copy_(self);
+
+  auto options = self.options().memory_format(LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  Tensor vals_ = at::empty_strided({n, 2}, {1, n}, options);
+  Tensor vecs_;
+  if (eigenvectors) {
+    vecs_ = at::empty_strided({n, n}, {1, n}, options);
+  } else {
+    vecs_ = at::empty({0}, options);
+  }
+
+  if (self.numel() != 0) {
+    int64_t info;
+    AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "eig_cpu", [&]{
+      apply_eig<scalar_t>(vals_, vecs_, self_, eigenvectors, &info);
+    });
+    singleCheckErrors(info, "eig_cpu");
+  }
+  return std::tuple<Tensor, Tensor>(vals_, vecs_);
+}
+
+std::tuple<Tensor&, Tensor&> eig_out_cpu(Tensor& vals, Tensor& vecs, const Tensor& self, bool eigenvectors) {
+  TORCH_CHECK(self.dtype() == vals.dtype(),
+    "Expected ", self.dtype(), " but got ", vals.dtype(), " dtype for eigenvalues output tensor");
+  TORCH_CHECK(!eigenvectors || (self.dtype() == vecs.dtype()),
+    "Expected ", self.dtype(), " but got ", vecs.dtype(), " dtype for eigenvectors output tensor");
+  Tensor tmp_vals, tmp_vecs;
+  std::tie(tmp_vals, tmp_vecs) = _eig_helper_cpu(self, eigenvectors);
```

Comment:
This function is, effectively, the only thing that's different between cpu and gpu implementation. Let's establish a good practice here, and use the same implementation of `eig` and `eig_out`, and register cpu and cuda dispatch for this helper function only. So in native_functions.yaml you'll have only `eig` and `eig_out`, dispatched on CPU, CUDA, in BatchLinearAlgebra.cpp there also will be only `eig` and `eig_out`, and you'd need to declare and register the correct dispatch for a helper. For a good example on how to do that look in `Cross.cpp` file, for a function `cross`.

Reply:
Ok, I've refactored so that CUDA and CPU share `eig` and `eig_out` implementations, and then dispatch to the specialized CPU or CUDA kernel